### DISCUSSION TITLE
(pipelines): add a way to add steps that aren't attached to stages

### DISCUSSION THREAD

**Discussion by markusl** [aws/aws-cdk#23888] [Q&A]
We receive an error "The given Stage construct ('foo/bar') should contain at least one Stack" when migrating our code to the modern pipelines.

### Reproduction Steps

```ts
pipeline.addStage(new cdk.Stage(scope, 'MyStage'));
```

### What did you expect to happen?

We would like to continue adding custom stages containing only CodeBuild projects and/or Lambda Invoke actions. Not all of our stages have stacks.

### Environment

  - **CDK CLI Version  :** 1.117.0 (build 0047c98)
  - **Framework Version:** 1.117.0 (build 0047c98)
  - **Node.js Version:** v14.15.0
  - **OS               :** macOS
  - **Language (Version):** TypeScript

### Workaround

```ts
  const stage = new cdk.Stage(scope, 'MyStage');
  new cdk.Stack(stage, 'EmptyStackRemoveThis', {
    stackName: 'EmptyStackRemoveThis',
  });
```

--- 

This is :bug: Bug Report

**Comment by rix0rrr** [comment]
`addStage` no longer means "add a CodePipeline Stage to the pipeline" -- it means: "deploy all stacks inside a `cdk.Stage`".

Adding arbitrary actions that have nothing to do with the deployment of CDK Stages is not strictly part of the design target, but I have a feeling using `addWave`, not adding any Stages to it and using `post` will do the same thing.

**Comment by markusl** [comment]
@rix0rrr thanks, `addWave` works indeed. 

So basically, you cannot do this:

```ts
pipeline.addStage(createInfra('dev', ...));
pipeline.addStage(createInfra('test', ...));
pipeline.addStage(createInfra('prod', ...));

// Does not deploy stacks but only runs some post-production actions (CodeBuild/Lambdas)

const postproductionStage = new cdk.Stage(scope, 'PostProduction');
// ... add actions

pipeline.addStage(postproductionStage);

```

Instead users can do this:

```ts
pipeline.addStage(createInfra('dev', ...));
pipeline.addStage(createInfra('test', ...));
pipeline.addStage(createInfra('prod', ...));

// Does not deploy stacks but only runs some post-production actions (CodeBuild/Lambdas)
const postproductionWave = pipeline.addWave('PostProduction');
postproductionWave.addPost(...);
```

I have to admit that it was not very intuitive to use `addWave` to add a CodePipeline stage instead of using `addStage` for adding a stage :) In general, the new API seems quite intuitive and easy to use but maybe more of these use cases could be clarified in the docs?

Br,
Markus

**Comment by rix0rrr** [comment]
It's unfortunate that "Stage" and "Stage" are called the same thing. The new API does not deal in terms of Pipeline Stages, it just happens to *create* them as an implementation detail as part of you interacting with it.

**Comment by nnboskovic** [comment]
I'm having a similar issue with CDK for C#. I have a stage that instantiates a stack with several components and I'm getting the ` Amazon.JSII.Runtime.JsiiException: The given Stage construct ('api/deploy') should contain at least one Stack` error. 

https://gist.github.com/nnbosko/5570135b3e72731a5e7bdb61f75829a9

**Comment by nnboskovic** [comment]
Still occurring in version 119.

**Comment by TheRealAmazonKendra** [comment]
It looks like the issue here has been addressed in the discussion here. So that this information continues to be easily searchable, I'm going to convert this into a discussion. If you believe this was done in error or are still experiencing problems related to this, please feel free to open a new issue.

**Reply by Outrun207** [reply]
Hello, any update on this? I don't see a circumstance where one would want to just synth and deploy code without linting/scanning/testing. It seems crucial to support adding stages that don't deploy stacks in this api.

**Comment by BELAGGOUN** [comment]
I'm facing the same issue ! I m not able to add a kind of the old  buildSpec.yaml where I can build a docker image run unit test and code test coverage ...etc . I have a feeling the the CI is completely ignored and the  CD is mixed with the infra code . any update on this issue please ?

**Comment by dennisvang** [comment]
@nnboskovic , @Outrun207, @BELAGGOUN 

Although not very intuitive, the workaround [suggested by @rix0rrr][1] does the job (also see [example by @markusl ][2]): Instead of a `cdk.Stage`, use a `cdk.Wave`. 

Here's a simplified example based on the [`WaveOptions` docs][3]:

```typescript
import * as pipelines from 'aws-cdk-lib/pipelines';

const myPipeline = new pipelines.CodePipeline(...);
const myCodeBuildStep = new pipelines.CodeBuildStep(...);

myPipeline.addWave('MyWave', {post: [myCodeBuildStep]});
```

[1]: https://github.com/aws/aws-cdk/discussions/23888#discussioncomment-4801825
[2]: https://github.com/aws/aws-cdk/discussions/23888#discussioncomment-4801826
[3]: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.WaveOptions.html
[4]: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#addwbrwaveid-options

**Comment by mlnrt** [comment]
I opened a feature request for this: https://github.com/aws/aws-cdk/issues/33011
